### ISSUE TITLE
Host model on Hugging Face

### ISSUE THREAD

**Issue by NielsRogge** [frank-xwang/UnSAM#5]
Hi folks,

Congrats on this great work! Would you be up for making your models available on the Hugging Face hub rather than Google Drive? This would make it easier available and discoverable for the community. Here's how to do that: https://huggingface.co/docs/hub/models-uploading.

Moreover, the models can be linked to the [paper](https://huggingface.co/papers/2407.02398), see [here](https://huggingface.co/docs/hub/en/model-cards#linking-a-paper) on how to do that.

Let me know if you need any help!

Kind regards,

Niels
ML Engineer @ HF

**Comment by frank-xwang** [comment]
Hi Niels, sounds like a good idea! Thank you! Will work on it after finishing the work on hand. :-)

**Comment by NielsRogge** [comment]
Awesome! Btw I noticed that the whole image segmentation models are [Mask2Former](https://huggingface.co/docs/transformers/en/model_doc/mask2former) checkpoints without any changes, correct?

In that case, they could be converted to the Hugging Face Transformers format so that people can use them using that API. The conversion script to convert the weights can be found here: https://github.com/huggingface/transformers/blob/main/src/transformers/models/mask2former/convert_mask2former_original_pytorch_checkpoint_to_pytorch.py.

**Comment by NielsRogge** [comment]
Hi @frank-xwang gentle ping on this :)

**Comment by cile98** [comment]
Hi @frank-xwang is there any progress or plans on doing this?

**Comment by yjwnb6** [comment]
@NielsRogge Hi Niels, I'm working on converting UnSAM to Hugging Face transformers. I rewrote `replace_swin_backbone `function(line 379) based on the attributes of ResNet50.[https://github.com/huggingface/transformers/blob/main/src/transformers/models/mask2former/convert_mask2former_original_pytorch_checkpoint_to_pytorch.py]
But the performance of the converted model is worse than the original UnSAM. While it generates approximately the same number of masks, it fails to catch some of the small objects. Could you take a look? Below are some examples of the difference of model performance. Thank you so much!

![WX20250104-215154@2x](https://github.com/user-attachments/assets/d4e68773-5ce1-4cc6-927a-fcaf9d086baa)
![WX20250104-215205@2x](https://github.com/user-attachments/assets/106a07bd-ee93-476d-a540-fdecd8610ae3)
![WX20250104-215227@2x](https://github.com/user-attachments/assets/3ccae1c5-71f1-4e66-b043-c14345eb97c0)
![WX20250104-215235@2x](https://github.com/user-attachments/assets/f2f53b1c-6949-4b6f-aefb-5ba4d535e76f)

**Comment by NielsRogge** [comment]
Hi, thanks for trying this out. Did you verify that logits match with the original implementation?

**Comment by yjwnb6** [comment]
> Hi, thanks for trying this out. Did you verify that logits match with the original implementation?

Yes, the logits match with the original model.

**Comment by NielsRogge** [comment]
If the logits match, it must be a difference in postprocessing. Did you check that they are equivalent? Things like threshold etc

**Comment by yjwnb6** [comment]
> If the logits match, it must be a difference in postprocessing. Did you check that they are equivalent? Things like threshold etc

Hi Niels, to generate better masks, I set `image_processor.do_resize = False` in line 985 of [convert_mask2former_original_pytorch_checkpoint_to_pytorch.py](https://github.com/huggingface/transformers/blob/main/src/transformers/models/mask2former/convert_mask2former_original_pytorch_checkpoint_to_pytorch.py). And I substitute line 883 - 886 with the following to avoid reshaping]

```
tr_complete = T.Compose(
    [T.ToTensor()],
)
y = (tr_complete(im) * 255.0).to(torch.int).float()
```
It turns out the logits shape didn't match after above modification. The logits match if I keep the original line 985 `image_processor.size = {"height": 384, "width": 384}`. Do you see why this is the case? I'm guessing the discrepancy in logits might be the issue. Thank you!

```
our_class_logits (HF model): torch.Size([1, 2000, 2])
original_class_logits (original UnSAM): torch.Size([1, 2000, 2])
our_mask_logits (HF model): torch.Size([1, 2000, 375, 563])
original_mask_logits (original UnSAM): torch.Size([1, 2000, 376, 568])
```

If I didn't do the resize modification, we have the matching logits
```
our_class_logits (HF model): torch.Size([1, 2000, 2])
original_class_logits (original UnSAM): torch.Size([1, 2000, 2])
our_mask_logits (HF model): torch.Size([1, 2000, 96, 96])
original_mask_logits (original UnSAM): torch.Size([1, 2000, 96, 96])
```

**Comment by NielsRogge** [comment]
Interesting! Pinging @qubvel here

**Comment by yujunwei04** [comment]
This issue has been resolved by PR: https://github.com/frank-xwang/UnSAM/pull/22

**Comment by NielsRogge** [comment]
Great! I've opened a PR for better metadata: https://huggingface.co/yujunwei04/unsam-whole-image-segmentation/discussions/1
